### PR TITLE
Convert actual realtime microservice from php to go

### PR DESCRIPTION
**Pull request by rbatistadev**
#### Type Of Change 
- [ ] Small bug fix
- [X] New feature or enhancement
- [ ] Breaking change

#### Checklist:
- [X] Commits are named and have tag following [commit rules](https://github.com/irontec/ivozprovider/blob/bleeding/doc/dev/en/commits.md)
- [X] Commits are split per component (schema, web/admin, kamusers, agis, ..)
- [X] Changes have been tested locally
- [ ] Fixes an existing issue (Fixes #XXXX) 
- [ ] Upport from existing Pull request #XXXX

#### Description
Convert actual realtime microservice from php to go

### CODE REVIEW

File: microservices/realtime/configs/config.yml [hunk 1]
```diff
@@ -0,0 +1,20 @@
+ws:
+   listen: 127.0.0.1:8081
+   cert: configs/server.crt
+   key: configs/server.key
+   jwt_public: /opt/irontec/ivozprovider/storage/jwt/public.pem
+   jwt_private: /opt/irontec/ivozprovider/storage/jwt/private.pem
+redis:
+   master: mymaster
```
Comment (mmadariaga) [inline microservices/realtime/configs/config.yml:8]: Shouldn't this be resolved by sentinel?

File: microservices/realtime/configs/config.yml [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+   sentinel_host: 127.0.0.1
+   sentinel_port: 26379
+   pool_size: 52
+   db: 1
+sentinels:
+   host: cache.ivozprovider.local
+   port: '26379'
```
Comment (mmadariaga) [inline microservices/realtime/configs/config.yml:15]: Are these diferent from sentinel_port and sentinel_host?

File: microservices/realtime/pkg/services/redis_status.go [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+
+var SIGNIFICANT_CALL_EVENTS = [...]string{CALL_SETUP, RINGING, IN_CALL, HANG_UP, UPDATE_CLID}
+
+func InitRedisControlClients() {
+	redisClient := redis.NewFailoverClient(&redis.FailoverOptions{
+		MasterName: config.GetRedisMaster(),
+		SentinelAddrs: []string{
+			config.GetRedisSentinelAddr(),
+		},
+		DB: 1,
```
Comment (mmadariaga) [inline microservices/realtime/pkg/services/redis_status.go:46]: read from config

File: microservices/realtime/pkg/services/redis_status.go [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+	ttl := func() int {
+		if event == "IN_CALL" {
+			return REDIS_KEYS_IN_CALL_TTL
+		}
+		return REDIS_KEYS_TTL
+	}()
+
+	redisClient.SetEx(
+		context.Background(),
+		channel, dataStr,
```
Comment (mmadariaga) [inline microservices/realtime/pkg/services/redis_status.go:125]:
```suggestion
		channel, 
		dataStr,
```

File: microservices/realtime/pkg/websocket/websocket_server.go [hunk 5]
```diff
@@ -0,0 +1,258 @@
...
+
+	storedClient, ok := redisPool.Load(conn)
+	if !ok {
+		fmt.Println("Could not find Redis client for the WebSocket connection.")
+		client := redis.NewFailoverClient(&redis.FailoverOptions{
+			MasterName: config.GetRedisMaster(),
+			SentinelAddrs: []string{
+				config.GetRedisSentinelAddr(),
+			},
+			DB: 1,
```
Comment (mmadariaga) [inline microservices/realtime/pkg/websocket/websocket_server.go:246]: duplicated  code?

**Review by mmadariaga — CHANGES_REQUESTED** [review]

**Comment by mmadariaga** [comment]
I miss old feed feature in scripts directory

File: microservices/realtime/configs/config.yml (hunk 1 above)
Comment (Kaian) [reply microservices/realtime/configs/config.yml:8]: Master name is `mymaster` by default, so propably can be skipped and the library will assume that value.

File: microservices/realtime/configs/config.yml (hunk 2 above)
Comment (Kaian) [reply microservices/realtime/configs/config.yml:15]: Looks like this sentinel section of the config is obsolete. The config package in the code reads the `redis.xxx` values

File: microservices/realtime/pkg/utils/constants.go [hunk 6]
```diff
@@ -0,0 +1,6 @@
+package utils
+
+// ROLES
+const ROLE_SUPER_ADMIN = "ROLE_SUPER_ADMIN"
+const ROLE_BRAND_ADMIN = "ROLE_BRAND_ADMIN"
+const ROLE_COMPANY_ADMIN = "ROLE_COMPANY_ADMIN"
```
Comment (Kaian) [inline microservices/realtime/pkg/utils/constants.go:6]: Are this constants used from more than one that package. If not, I will move it to the package that use them.
The other alternative is to move all constants here, like redis_status.go constants

File: microservices/realtime/server
Comment (Kaian) [inline microservices/realtime/server:1]: Binary files should not be commited

File: microservices/realtime/tests/test-build.sh [hunk 7]
```diff
@@ -0,0 +1,11 @@
...
+
+set -xe
+
+cd $(dirname $(realpath $0))/..
+
+# Download required dependencies
+go mod download && go mod verify
+
+# Build server
+go build -v -o deitu-server irontec.com/realtime/cmd/server
```
Comment (Kaian) [inline microservices/realtime/tests/test-build.sh:11]: what is deitu?

File: microservices/realtime/tests/test-codestyle.sh [hunk 8]
```diff
@@ -0,0 +1,15 @@
...
+go fmt ./...
+
+# Check there is updated files
+if [ -n "$(git status -s | grep .go)" ]; then
+    echo "Some files have incorrect format."
+    git status -s | grep .go
+    exit 1
+fi
+
+exit 0
```
Comment (Kaian) [inline microservices/realtime/tests/test-codestyle.sh:15]: We should add a pipeline step for realtime codestyle + build tests

File: microservices/realtime/configs/config.yml [hunk 9]
```diff
@@ -0,0 +1,20 @@
+ws:
+   listen: 127.0.0.1:8081
+   cert: configs/server.crt
+   key: configs/server.key
```
Comment (Kaian) [inline microservices/realtime/configs/config.yml:4]: Cert and key are not required. This is an HTTP server

File: microservices/realtime/configs/config.yml [hunk 10]
```diff
@@ -0,0 +1,20 @@
...
+   jwt_public: /opt/irontec/ivozprovider/storage/jwt/public.pem
+   jwt_private: /opt/irontec/ivozprovider/storage/jwt/private.pem
+redis:
+   master: mymaster
+   sentinel_host: 127.0.0.1
```
Comment (Kaian) [inline microservices/realtime/configs/config.yml:9]: sentinel_host must be set to `cache.ivozprovider.local`.

File: microservices/realtime/configs/config.yml [hunk 11]
```diff
@@ -0,0 +1,20 @@
...
+   sentinel_port: 26379
+   pool_size: 52
```
Comment (Kaian) [inline microservices/realtime/configs/config.yml:11]: Not sure what `pool_size` is used for

File: microservices/realtime/pkg/utils/httputil.go [hunk 12]
```diff
@@ -0,0 +1,18 @@
...
+		Message: err.Error(),
+	}
+	ctx.JSON(status, er)
+}
+
+// HTTPError example
+type HTTPError struct {
+	Code    int    `json:"code" example:"400"`
+	Message string `json:"message" example:"status bad request"`
+}
```
Comment (Kaian) [inline microservices/realtime/pkg/utils/httputil.go:18]: Where is this used? 
Do you require HTTP response generators for anything?
Or gin-gonic?

File: microservices/realtime/pkg/services/http.go [hunk 13]
```diff
@@ -0,0 +1,86 @@
...
+	case utils.ROLE_BRAND_ADMIN:
+		api = config.GetHttpApiBrand()
+	case utils.ROLE_COMPANY_ADMIN:
+		api = config.GetHttpApiClient()
+	}
+
+	apiUrl := fmt.Sprintf(
+		"%s/my/active_calls/realtime_filter",
+		api,
+	)
```
Comment (Kaian) [inline microservices/realtime/pkg/services/http.go:40]: This uses an endpoint that does not exist in any API. Should we add here the commits required to make it work?

Comment (rbatistadev) [reply microservices/realtime/pkg/services/http.go:40]: It had been designed to add a future branch, which already exists with that endpoint.

File: debian/ivozprovider-realtime.install [hunk 14]
```diff
@@ -1 +1,2 @@
 microservices/realtime /opt/irontec/ivozprovider/microservices/
+microservices/realtime/configs/config.yml
```
Comment (Kaian) [inline debian/ivozprovider-realtime.install:2]: Missing destination folder, format of lines in this file is `src dst`

File: microservices/realtime/pkg/services/feeder/call.go [hunk 15]
```diff
@@ -0,0 +1,121 @@
...
+
+// GetStatus returns the call status.
+func (c *AbstractCall) GetStatus() string {
+	return c.Status
+}
+
+// GetChannel returns the call channel.
+func (c *AbstractCall) GetChannel() string {
+	return c.Channel
+}
```
Comment (Kaian) [inline microservices/realtime/pkg/services/feeder/call.go:46]: If you add getters for struct fields, it's probably a good idea to make fields private to enforce the usage of these getters

File: microservices/realtime/scripts/build.sh [hunk 16]
```diff
@@ -0,0 +1,4 @@
+cd $(dirname $(realpath $0))/..
```
Comment (Kaian) [inline microservices/realtime/scripts/build.sh:1]: Missing shebang `#!/bin/bash`

**Review by Kaian — CHANGES_REQUESTED** [review]

**Review by Kaian — APPROVED** [review]

**Review by mmadariaga — APPROVED** [review]

**Pull request merged**